spec.md

```diff
@@ -1601,6 +1655,72 @@ The CO MUST implement the specified error recovery behavior when it encounters t
+
+This call can be made by the CO during any time in the lifecycle of the volume after creation if plugin has `EXPANDABLE_ONLINE` capability.
+If plugin has `EXPAND_VOLUME` node capability, then `NodeExpandVolume` MUST be called after `ControllerExpandVolume` has completed.
+
+
+`ControllerExpandVolume` MUST be called after `ControllerUnpublishVolume` if plugin has only `EXPANDABLE_OFFLINE` capability.
+A plugin may return `FAILED_PRECONDITION` error if underlying volume can not be expanded when published to a node.
+
+`ControllerExpandVolume` RPC call MUST be idempotent and if size of underlying volume already meets requested capacity, the plugin MUST respond with successfull response.
+
+Examples:
+* Offline Volume Expansion:
+  Given a ElasticSearch process that runs on Azure Disk volume and needs more space.
+  - The administrator takes the elasticsearch server offline and request more space for volume from the CO.
+  - The CO in turn first makes `ControllerExpandVolume` RPC call which results in requesting more space from Azure cloudprovider for volume-id that was being used by ElasticSearch.
+  - Once `ControllerExpandVolume` finishes the CO will inform administrator about it and administrator will resume the ElasticSearch workload.
+  - On node where ElasticSearch workload is scheduled when workload starts after `NodeStageVolume` the CO calls `NodeExpandVolume`.
+  - Calling `NodeExpandVolume` on volume results in expanding the underlying file system and added space becomes available to workload when it startsup.
+* Online Volume Expansion:
+  Given a Mysql server running using Openstack Cinder and needs more space.
+  - The administrator requests more space for volume from the CO.
+  - The CO in turn first makes `ControllerExpandVolume` RPC call which results in requesting more space from Openstack Cinder for given volume.
+  - The node where mysql workload is running the CO makes `NodeExpandVolume` RPC call while volume is in-use using path where volume is staged.
+  - Calling `NodeExpandVolume` on volume results in expanding the underlying file system and added space automatically becomes available to mysql workload without any downtime.
+
+
+```protobuf
+message ControllerExpandVolumeRequest {
+  // The ID of the volume to expand. This field is REQUIRED.
+  string volume_id = 1;
+
+  // This field is REQUIRED. This allows CO to specify the
+  // capacity requirements of the volume after expansion.
+  CapacityRange capacity_range = 2;
+
+  // Secrets required by the plugin for expanding the volume.
+  // This field is OPTIONAL.
+  map<string, string> controller_expand_volume_secrets = 3;
+}
+
+message ControllerExpandVolumeResponse {
+  // Capacity of volume after expansion.
+  // This field is REQUIRED.
+  int64 capacity_bytes = 1;
+
+  // Whether file system expansion is required for the volume.
+  // When true the CO MUST make NodeExpandVolume RPC call
+  // on the node. This field is REQUIRED.
+  bool fs_resize_required = 2;
```

Comment:
I am wondering if this field is necessary. Can we leverage the EXPAND_VOLUME node capability to inform the CO to call `NodeExpandVolume`?

Reply:
It is a convenience field and has few benefits:

a. Say you created 5 GB volume on GCE for mysql but before you could start the workload you realized that is not going to be enough. So you edit the volume(pvc in k8s for example) and request more space, the plugin could possibly determine that no `NodeExpandVolume` is needed in this case because volume has never been formatted and return false. 
b. It could save control plane from querying node capabilities which is useful sometimes (cc @saad-ali ).